src/transaction/log_global.c

```diff
@@ -102,29 +106,41 @@ log_global::~log_global ()
   delete writer_info;
 }
 
+#if defined (SERVER_MODE)
 void
-log_global::update_max_ps_flushed_lsa (const LOG_LSA &lsa)
+log_global::wakeup_ps_flush_waiters ()
 {
-  {
-    std::unique_lock<std::mutex> lock (m_ps_lsa_mutex);
-    m_max_ps_flushed_lsa = lsa;
-  } // lock.unlock ();
+  assert (is_active_transaction_server ());
+  m_ps_lsa_up_to_date.store (false);
   m_ps_lsa_cv.notify_all ();
 }
 
 void
-log_global::wait_flushed_lsa (const log_lsa &flush_lsa)
+log_global::wait_for_ps_flushed_lsa (const log_lsa &flush_lsa)
 {
-  if (m_max_ps_flushed_lsa >= flush_lsa)
+  std::unique_lock<std::mutex> ulock (m_ps_lsa_mutex);
+  while (m_ps_consensus_flushed_lsa < flush_lsa)
     {
-      // already flushed
-      return;
+      // Only who first notices it's out-dated updates the consensus lsa.
+      if (!m_ps_lsa_up_to_date.exchange (true)) 
```

Comment:
Your approach would work correctly, but some waste can occur when the condition is not met:
- The saved_lsa is received, so It wakes up all transactions. However, no transaction is happy with the newly calculated consensus lsa. It makes all transactions do the same calculations and sleep again.
- Every new transaction starts waiting and calculates the LSA if the LSA hasn't been met for the transaction.

To avoid the useless calculation, which can occur whenever a saved_lsa is received, the condition variable is needed. Do you think it's better to ignore it?

> might even be a good idea to switch to a `condition_variable::wait_for` instead of

Do we have something to do periodically in the middle of waiting? Or is it for the future like timeout?

Reply:
> The saved_lsa is received, so It wakes up all transactions. However, no transaction is happy with the newly calculated consensus lsa. It makes all transactions do the same calculations and sleep again.

The way I see it:
- all transactions that are waiting are doing so by the `m_ps_lsa_cv` and guarded by `m_ps_lsa_mutex`
- and  only one transaction is able to acquire the mutex
- and execute the condition lambda body:
  - check value
  - if not happy with the calculated value, do the calculation
  - then check value again
    - either happy and return
      - after unlocking the unique lock, it should notify again all other potentially waiting transactions which will then first check the up-to-date calculated value and return or do the calculation again (indeed here is the potential that each transaction that is unhappy with the value will end up calculating the value again, but I see this as more of a corner case)
    - or not happy and continue waiting

Let's leave the code as you proposed and see how it behaves.
We'll change it later with another refactoring if needed.
My idea was to avoid introducing yet another flag variable, but maybe it is forced. :)